wazuh/wazuh
# Implementing self-recovery mechanism for RocksDB corruption

|Related issue|
|---|
|Closes #24151|



## Description



Work in progress.

This PR:
- Adds a more verbose logging option for RocksDB to give more details in case of failure.
- The logger helper class was updated because the utils methods lack of logging.
- Executes a repair method by default when the opening fails due to corruption (for queues and wrapper). It also prints a warning message after a successful recovery. This mechanism is optional in the wrapper and can be disabled in the constructor

Extra changes:
- A retry mechanism was added for the indexer connector test because it was flaky: https://github.com/wazuh/wazuh/pull/24333/commits/011a1c42d51fe1cbd6ff31f1b9ea6b9ffe2aa104

> [!NOTE]  
> Now that all the DBs are repaired by default, the opening of the `feed` database disables the automatic repair in the wrapper constructor.  The re-download mechanism that was working before these changes is triggered. 

> [!NOTE]  
> Considering the approach taken to supply a tag to the helper without changing the firm of all the intermediate methods, the tag of the utils methods will always be the same: `logger-helper`.


## Tests



**Manual tests**

We can find the following RocksDB databases in a common installation, with the indexer enabled and the content downloaded

```bash
root@fb659c773250:/var/ossec# find . -name "MANIFEST-*"              
./queue/keystore/MANIFEST-000014
./queue/vd/state_track/MANIFEST-000005
./queue/vd/feed/MANIFEST-001740
./queue/vd/event/MANIFEST-000005
./queue/vd/delayed/MANIFEST-000009
./queue/vd/reports/MANIFEST-000005
./queue/vd/inventory/MANIFEST-000015
./queue/indexer/wazuh-states-vulnerabilities-fb659c773250/MANIFEST-000005
./queue/indexer/db/wazuh-states-vulnerabilities-fb659c773250/MANIFEST-000005
./queue/vd_updater/rocksdb/updater_vulnerability_feed_manager_metadata/MANIFEST-000014
```

We can force a corruption error by removing the `MANIFEST` file and then restarting the manager to make sure the module can recover and continue working.


<details><summary>Breaking all DBs at once except feed</summary>
<p>

We can find all `MANIFEST` files with this command, and I removed all except the one from `feed` DB

```
find . -name "MANIFEST-*" 
```

Then, we find that the module repairs the DBs and start succesfully

```
2024/07/20 04:18:32 logger-helper: WARNING: Database 'queue/vd/state_track' was repaired because it was corrupt.
2024/07/20 04:18:32 logger-helper: WARNING: Database 'queue/indexer/db/wazuh-states-vulnerabilities-fb659c773250' was repaired because it was corrupt.
2024/07/20 04:18:32 logger-helper: WARNING: Database 'queue/keystore' was repaired because it was corrupt.
2024/07/20 04:18:32 logger-helper: WARNING: Database 'queue/indexer/wazuh-states-vulnerabilities-fb659c773250' was repaired because it was corrupt.
2024/07/20 04:18:32 logger-helper: WARNING: Database 'queue/vd/reports' was repaired because it was corrupt.
2024/07/20 04:18:32 logger-helper: WARNING: Database 'queue/vd/event' was repaired because it was corrupt.
2024/07/20 04:18:33 logger-helper: WARNING: Database 'queue/vd_updater/rocksdb/updater_vulnerability_feed_manager_metadata' was repaired because it was corrupt.
2024/07/20 04:18:33 logger-helper: WARNING: Database 'queue/vd/inventory' was repaired because it was corrupt.
2024/07/20 04:18:33 logger-helper: WARNING: Database 'queue/vd/delayed' was repaired because it was corrupt.
```


</p>
</details> 

<details><summary>Breaking feed DB</summary>
<p>

When we corrupt the `feed` DB, the fallback mechanism is triggered

```
2024/07/20 04:25:13 wazuh-modulesd:vulnerability-scanner: ERROR: Error opening the database: Failed to list columns, 'queue/vd/feed' won't be repaired: While opening a file for sequentially reading: queue/vd/feed/MANIFEST-001788: No such file or directory, trying to re-download the feed.
2024/07/20 04:25:13 wazuh-modulesd:vulnerability-scanner: INFO: Vulnerability scanner module started.
2024/07/20 04:25:40 wazuh-modulesd:vulnerability-scanner: INFO: Initiating update feed process.
```

Consider that:
- The `queue/vd/feed/` isn't repaired because we can't add offsets to a damaged DB . In case of failure, it is removed and a re-download is triggered.

</p>
</details> 


**Checks**


- Compilation without warnings in every supported platform
- [x] Added unit tests (for new features)
